x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/saved_objects_adapter/saved_objects_adapter.ts

```diff
@@ -0,0 +1,187 @@
+      bulkCreateTimeDurations: [],
+      searchAfterTimeDurations: [],
+      lastLookBackDate: null,
+    });
+  };
+
+  private getOrCreateRuleStatuses = async (
+    ruleId: string
+  ): Promise<Array<SavedObject<IRuleStatusSOAttributes>>> => {
+    const ruleStatuses = await this.find({
+      spaceId: '',
+      ruleId,
+      logsCount: MAX_RULE_STATUSES,
+    });
+    if (ruleStatuses.length > 0) {
+      return ruleStatuses;
+    }
+    const newStatus = await this.createNewRuleStatus(ruleId);
+
+    return [newStatus];
+  };
+
+  public async logStatusChange({ newStatus, ruleId, message, metrics }: LogStatusChangeArgs) {
+    switch (newStatus) {
+      case RuleExecutionStatus['going to run']:
+      case RuleExecutionStatus.succeeded:
+      case RuleExecutionStatus.warning:
+      case RuleExecutionStatus['partial failure']: {
+        const [currentStatus] = await this.getOrCreateRuleStatuses(ruleId);
+
+        await this.ruleStatusClient.update(currentStatus.id, {
+          ...currentStatus.attributes,
+          ...buildRuleStatusAttributes(newStatus, message, metrics),
+        });
+
+        break;
+      }
+
+      case RuleExecutionStatus.failed: {
+        const ruleStatuses = await this.getOrCreateRuleStatuses(ruleId);
+        const [currentStatus] = ruleStatuses;
+
+        const failureAttributes = {
+          ...currentStatus.attributes,
+          ...buildRuleStatusAttributes(RuleExecutionStatus.failed, message, metrics),
+        };
+
+        // We always update the newest status, so to 'persist' a failure we push a copy to the head of the list
+        await this.ruleStatusClient.update(currentStatus.id, failureAttributes);
+        const lastStatus = await this.ruleStatusClient.create(failureAttributes);
+
+        // drop oldest failures
+        const oldStatuses = [lastStatus, ...ruleStatuses].slice(MAX_RULE_STATUSES);
+        await Promise.all(oldStatuses.map((status) => this.delete(status.id)));
+      }
+    }
+  }
+}
+
+const buildRuleStatusAttributes: (
+  status: RuleExecutionStatus,
+  message?: string,
+  metrics?: LegacyMetrics
+) => Partial<IRuleStatusSOAttributes> = (status, message, metrics = {}) => {
+  const now = new Date().toISOString();
+  const baseAttributes: Partial<IRuleStatusSOAttributes> = {
+    ...metrics,
+    status:
+      status === RuleExecutionStatus.warning ? RuleExecutionStatus['partial failure'] : status,
+    statusDate: now,
+  };
+
+  switch (status) {
+    case RuleExecutionStatus.succeeded:
+    case RuleExecutionStatus.warning:
+    case RuleExecutionStatus['partial failure']: {
+      return {
+        ...baseAttributes,
+        lastSuccessAt: now,
+        lastSuccessMessage: message,
+      };
+    }
+    case RuleExecutionStatus.failed: {
+      return {
+        ...baseAttributes,
+        lastFailureAt: now,
+        lastFailureMessage: message,
+      };
+    }
+    case RuleExecutionStatus['going to run']: {
+      return baseAttributes;
+    }
+  }
```

Comment:
In the rule status service this function had `assertUnreachable(status);` at the and, is it possible to get it back?

I'd add it to `logStatusChange` too.

Reply:
`assertUnreachable` is redundant in this case. As the return type is specified, Typescript performs an exhaustiveness check and warns us if we forgot something. But for `logStatusChange` it could be helpful.

Reply:
Added `assertUnreachable` to `logStatusChange`